### ISSUE TITLE
HAP server error: Dictionary changed size during iteration

### ISSUE THREAD

**Issue by cdce8p** [ikalchev/HAP-python#172]
Error reported by @ehendrix23
Observed with `2.4.0`, but as the diff suggested, it hasn't been fixed yet.

```
2018-11-12 08:42:48 INFO (SyncWorker_0) [pyhap.hap_server] Stopping HAP server
----------------------------------------
Exception happened during processing of request from ('xxx.xxx.x.xxx', 51818)
2018-11-12 08:42:48 ERROR (SyncWorker_18) [asyncio] Task exception was never retrieved future:
  <Task finished coro=<AccessoryDriver.async_stop() done,
  defined at /Users/Media/homeassistant/lib/python3.7/site-packages/pyhap/accessory_driver.py:278>
  exception=RuntimeError('dictionary changed size during iteration')>
Traceback (most recent call last):
File "/Users/Media/homeassistant/lib/python3.7/site-packages/pyhap/accessory_driver.py", line 280, in async_stop
  await self.async_add_job(self._do_stop)
File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/concurrent/futures/thread.py", line 57, in run
  result = self.fn(*self.args, **self.kwargs)
File "/Users/Media/homeassistant/lib/python3.7/site-packages/pyhap/accessory_driver.py", line 310, in _do_stop
  self.http_server.server_close()
File "/Users/Media/homeassistant/lib/python3.7/site-packages/pyhap/hap_server.py", line 830, in server_close
  for sock in self.connections.values():
RuntimeError: dictionary changed size during iteration
```

**Comment by ikalchev** [comment]
This looks like a race between the server close and _close_socket. I think I know how to fix it, but I will try to reproduce first.

@ehendrix23, does this happen every time you stop HAP-python?

**Comment by ehendrix23** [comment]
@ikalchev , just checked. Saw it twice and that's it. Both would have been with 2.4.0.

**Comment by cdce8p** [comment]
@ikalchev Any update?

**Comment by ikalchev** [comment]
Can you bump to 2.4.1 in the meantime?

I haven’t managed to test my fix fully yet, I wil most probably do it over the weekend.

**Comment by cdce8p** [comment]
Yeah, no Problem. Just pushed the update to Home Assistant: https://github.com/home-assistant/home-assistant/pull/18417

@ehendrix23 Have you tested 2.4.1 yet?

**Comment by ehendrix23** [comment]
@ikalchev , @cdce8p , I have been running on 2.4.1 since monday morning when cdce8p asked me to. :-)

I just checked back through my log files and I only have the 2 occurrences of this when I was on 2.4.0 happened when stopping HASS).
Since then (just counted) I restarted HASS 41 times (update configurations for other stuff, test some of my new code, ...) and at no time has that error been reported anymore on 2.4.1.

Also note that stopping HASS (and thus HAP server) now has always been very quick. No issue anymore with it hanging. :-) 
Just checked as well and I also do not see any in CLOSE_WAIT (right it now has been up and running for 16 hours).
